quicklens/src/main/scala-3/com.softwaremill.quicklens/package.scala

```diff
@@ -0,0 +1,403 @@
+    def atOrElse(default: => T): T = ???
+
+    @compileTimeOnly(canOnlyBeUsedInsideModify("index"))
+    def index: T = ???
+
+  extension [A](value: A)
+    @compileTimeOnly(canOnlyBeUsedInsideModify("when"))
+    def when[B <: A]: B = ???
+
+  extension [T, U](f1: T => PathModify[T, U])
+    def andThenModify[V](f2: U => PathModify[U, V]): T => PathModify[T, V] = (t: T) =>
+      PathModify[T, V](t, vv => f1(t).f(u => f2(u).f(vv)))
+
+  private def canOnlyBeUsedInsideModify(method: String) = s"$method can only be used as a path component inside modify"
+
+  //
+
+  def toPathModify[S: Type, A: Type](obj: Expr[S], f: Expr[(A => A) => S])(using Quotes): Expr[PathModify[S, A]] = '{ PathModify( ${obj}, ${f} ) }
+
+  def fromPathModify[S: Type, A: Type](pathModify: Expr[PathModify[S, A]])(using Quotes): Expr[(A => A) => S] = '{ ${pathModify}.f }
+
+  def to[T: Type, R: Type](f: Expr[T] => Expr[R])(using Quotes): Expr[T => R] = '{ (x: T) => ${ f('x) } }
+
+  def from[T: Type, R: Type](f: Expr[T => R])(using Quotes): Expr[T] => Expr[R] = (x: Expr[T]) => '{ $f($x) }
+
+  def modifyAllImpl[S, A](obj: Expr[S], focus: Expr[S => A], focusesExpr: Expr[Seq[S => A]])(using qctx: Quotes, tpeS: Type[S], tpeA: Type[A]): Expr[PathModify[S, A]] = {
+    import qctx.reflect.*
+
+    val focuses = focusesExpr match {
+      case Varargs(args) => args
+    }
+
+    val modF1 = fromPathModify(modifyImpl(obj, focus))
+    val modF = to[(A => A), S] { (mod: Expr[A => A]) =>
+      focuses.foldLeft(from[(A => A), S](modF1).apply(mod)) {
+        case (objAcc, focus) =>
+          val modCur = fromPathModify(modifyImpl(objAcc, focus))
+          from[(A => A), S](modCur).apply(mod)
+      }
+    }
+
+    toPathModify(obj, modF)
+  }
+
+  def modifyImpl[S, A](obj: Expr[S], focus: Expr[S => A])(using qctx: Quotes, tpeS: Type[S], tpeA: Type[A]): Expr[PathModify[S, A]] = {
+    import qctx.reflect.*
+    
+    def unsupportedShapeInfo(tree: Tree) = s"Unsupported path element. Path must have shape: _.field1.field2.each.field3.(...), got: $tree"
+
+    def methodSupported(method: String) =
+      Seq("at", "each", "eachWhere", "eachRight", "eachLeft", "atOrElse", "index").contains(method)
+
+    enum PathSymbol:
+      case Field(name: String)
+      case FunctionDelegate(name: String, givn: Term, typeTree: TypeTree, args: List[Term])
+
+    def toPath(tree: Tree): Seq[PathSymbol] = {
+      tree match {
+        /** Field access */
+        case Select(deep, ident) =>
+          toPath(deep) :+ PathSymbol.Field(ident)
+        /** Method call with arguments and using instance */
+        case Apply(Apply(Apply(TypeApply(Ident(s), typeTrees), idents), args), List(givn)) if methodSupported(s) =>
+          idents.flatMap(toPath) :+ PathSymbol.FunctionDelegate(s, givn, typeTrees.last, args)
+        /** Method call with no arguments and using instance */
+        case a@Apply(Apply(TypeApply(Ident(s), typeTrees), idents), List(givn)) if methodSupported(s) =>
```

Comment:
maybe it would make sense to split the `methodSupported` into two sets - one-arg and no-arg methods - what do you think? that could be more precise